The job **`aarch64-gnu-llvm-19-1`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/17067068913/job/48386707491) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/48386707491) [(plain)](https://github.com/rust-lang/rust/commit/a68e8bee7efaf1d0000a72fe4e348fd9834a419e/checks/48386707491/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
failures:

---- [codegen] tests/codegen-llvm/function-arguments.rs stdout ----

error: verification with 'FileCheck' failed
status: exit status: 1
command: "/usr/lib/llvm-19/bin/FileCheck" "--input-file" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/codegen-llvm/function-arguments/function-arguments.ll" "/checkout/tests/codegen-llvm/function-arguments.rs" "--check-prefix=CHECK" "--allow-unused-prefixes" "--dump-input-context" "100"
stdout: none
--- stderr -------------------------------
/checkout/tests/codegen-llvm/function-arguments.rs:83:11: error: CHECK: expected string not found in input
// CHECK: @readonly_borrow(ptr noalias noundef readonly align 4{{( captures\(address, read_provenance\))}} dereferenceable(4) %_1)
          ^
/checkout/obj/build/aarch64-unknown-linux-gnu/test/codegen-llvm/function-arguments/function-arguments.ll:320:55: note: scanning from here
define noundef i64 @option_nonzero_int(i64 noundef %x) unnamed_addr #1 {
                                                      ^

Input file: /checkout/obj/build/aarch64-unknown-linux-gnu/test/codegen-llvm/function-arguments/function-arguments.ll
Check file: /checkout/tests/codegen-llvm/function-arguments.rs

-dump-input=help explains the following input dump.

Input was:
<<<<<<
          .
          .
          .
        220:  %ptr.1 = load ptr, ptr %2, align 8, !nonnull !2, !align !4, !noundef !2 
        221:  %3 = icmp ne ptr %ptr.0, null 
        222:  call void @llvm.assume(i1 %3) 
        223:  %4 = icmp ne ptr %ptr.0, null 
        224:  call void @llvm.assume(i1 %4) 
        225:  call void @llvm.lifetime.start.p0(i64 8, ptr %1) 
        226:  %5 = getelementptr inbounds i8, ptr %ptr.1, i64 8 
        227:  %6 = load i64, ptr %5, align 8, !range !6, !invariant.load !2 
        228:  %7 = getelementptr inbounds i8, ptr %ptr.1, i64 16 
        229:  %8 = load i64, ptr %7, align 8, !range !7, !invariant.load !2 
        230:  store i64 %6, ptr %1, align 8 
        231:  %size = load i64, ptr %1, align 8, !noundef !2 
        232:  call void @llvm.lifetime.end.p0(i64 8, ptr %1) 
        233:  call void @llvm.lifetime.start.p0(i64 8, ptr %0) 
        234:  %9 = getelementptr inbounds i8, ptr %ptr.1, i64 8 
        235:  %10 = load i64, ptr %9, align 8, !range !6, !invariant.load !2 
        236:  %11 = getelementptr inbounds i8, ptr %ptr.1, i64 16 
        237:  %12 = load i64, ptr %11, align 8, !range !7, !invariant.load !2 
        238:  store i64 %12, ptr %0, align 8 
        239:  %align = load i64, ptr %0, align 8, !noundef !2 
        240:  call void @llvm.lifetime.end.p0(i64 8, ptr %0) 
        241:  br label %bb7 
        242:  
        243: bb7: ; preds = %start 
        244:  %13 = sub i64 %align, 1 
        245:  %14 = icmp ule i64 %13, 9223372036854775807 
        246:  call void @llvm.assume(i1 %14) 
        247:  %15 = getelementptr inbounds i8, ptr %layout, i64 8 
        248:  store i64 %size, ptr %15, align 8 
        249:  store i64 %align, ptr %layout, align 8 
        250:  %16 = icmp eq i64 %size, 0 
        251:  br i1 %16, label %bb3, label %bb1 
        252:  
        253: bb3: ; preds = %bb1, %bb7 
        254:  ret void 
        255:  
        256: bb1: ; preds = %bb7 
        257:  %_7 = getelementptr inbounds i8, ptr %self, i64 16 
        258:  %17 = icmp ne ptr %ptr.0, null 
        259:  call void @llvm.assume(i1 %17) 
        260:  %18 = load i64, ptr %layout, align 8, !range !5, !noundef !2 
        261:  %19 = getelementptr inbounds i8, ptr %layout, i64 8 
        262:  %20 = load i64, ptr %19, align 8, !noundef !2 
        263: ; call <alloc::alloc::Global as core::alloc::Allocator>::deallocate 
        264:  call void @"_ZN63_$LT$alloc..alloc..Global$u20$as$u20$core..alloc..Allocator$GT$10deallocate17he6fcd87b9c0c3dc6E"(ptr noalias noundef nonnull readonly align 1 %_7, ptr noundef nonnull %ptr.0, i64 noundef %18, i64 noundef %20) 
        265:  br label %bb3 
        266:  
        267: bb6: ; No predecessors! 
        268:  unreachable 
        269: } 
        270:  
        271: ; Function Attrs: uwtable 
        272: define noundef zeroext i1 @boolean(i1 noundef zeroext %x) unnamed_addr #1 { 
        273: start: 
        274:  ret i1 %x 
        275: } 
        276:  
        277: ; Function Attrs: uwtable 
        278: define i8 @maybeuninit_boolean(i8 %x) unnamed_addr #1 { 
        279: start: 
        280:  ret i8 %x 
        281: } 
        282:  
        283: ; Function Attrs: uwtable 
        284: define noundef zeroext i1 @enum_bool(i1 noundef zeroext %x) unnamed_addr #1 { 
        285: start: 
        286:  ret i1 %x 
        287: } 
        288:  
        289: ; Function Attrs: uwtable 
        290: define i8 @maybeuninit_enum_bool(i8 %x) unnamed_addr #1 { 
        291: start: 
        292:  ret i8 %x 
        293: } 
        294:  
        295: ; Function Attrs: uwtable 
        296: define noundef range(i32 0, 1114112) i32 @char(i32 noundef range(i32 0, 1114112) %x) unnamed_addr #1 { 
        297: start: 
        298:  ret i32 %x 
        299: } 
        300:  
        301: ; Function Attrs: uwtable 
        302: define i32 @maybeuninit_char(i32 %x) unnamed_addr #1 { 
        303: start: 
        304:  ret i32 %x 
        305: } 
        306:  
        307: ; Function Attrs: uwtable 
        308: define noundef i64 @int(i64 noundef %x) unnamed_addr #1 { 
        309: start: 
        310:  ret i64 %x 
        311: } 
        312:  
        313: ; Function Attrs: uwtable 
        314: define noundef range(i64 1, 0) i64 @nonzero_int(i64 noundef range(i64 1, 0) %x) unnamed_addr #1 { 
        315: start: 
        316:  ret i64 %x 
        317: } 
        318:  
        319: ; Function Attrs: uwtable 
        320: define noundef i64 @option_nonzero_int(i64 noundef %x) unnamed_addr #1 { 
check:83                                                           X~~~~~~~~~~~~~~~~~~ error: no match found
        321: start: 
check:83     ~~~~~~~
        322:  ret i64 %x 
check:83     ~~~~~~~~~~~~
        323: } 
check:83     ~~
        324:  
check:83     ~
        325: ; Function Attrs: uwtable 
check:83     ~~~~~~~~~~~~~~~~~~~~~~~~~~
        326: define void @readonly_borrow(ptr noalias noundef readonly align 4 dereferenceable(4) %_1) unnamed_addr #1 { 
check:83     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        327: start: 
check:83     ~~~~~~~
        328:  ret void 
check:83     ~~~~~~~~~~
        329: } 
check:83     ~~
        330:  
check:83     ~
        331: ; Function Attrs: uwtable 
check:83     ~~~~~~~~~~~~~~~~~~~~~~~~~~
        332: define noundef align 4 dereferenceable(4) ptr @readonly_borrow_ret() unnamed_addr #1 { 
check:83     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        333: start: 
check:83     ~~~~~~~
        334:  br label %bb1 
check:83     ~~~~~~~~~~~~~~~
        335:  
check:83     ~
        336: bb1: ; preds = %bb1, %start 
check:83     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        337:  br label %bb1 
check:83     ~~~~~~~~~~~~~~~
        338: } 
check:83     ~~
        339:  
check:83     ~
        340: ; Function Attrs: uwtable 
check:83     ~~~~~~~~~~~~~~~~~~~~~~~~~~
        341: define void @static_borrow(ptr noalias noundef readonly align 4 dereferenceable(4) %_1) unnamed_addr #1 { 
check:83     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        342: start: 
check:83     ~~~~~~~
        343:  ret void 
check:83     ~~~~~~~~~~
        344: } 
check:83     ~~
        345:  
check:83     ~
        346: ; Function Attrs: uwtable 
check:83     ~~~~~~~~~~~~~~~~~~~~~~~~~~
        347: define void @named_borrow(ptr noalias noundef readonly align 4 dereferenceable(4) %_1) unnamed_addr #1 { 
check:83     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        348: start: 
check:83     ~~~~~~~
        349:  ret void 
check:83     ~~~~~~~~~~
        350: } 
check:83     ~~
        351:  
check:83     ~
        352: ; Function Attrs: uwtable 
check:83     ~~~~~~~~~~~~~~~~~~~~~~~~~~
        353: define void @unsafe_borrow(ptr noundef nonnull align 2 %_1) unnamed_addr #1 { 
check:83     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        354: start: 
check:83     ~~~~~~~
        355:  ret void 
check:83     ~~~~~~~~~~
        356: } 
check:83     ~~
        357:  
check:83     ~
        358: ; Function Attrs: uwtable 
check:83     ~~~~~~~~~~~~~~~~~~~~~~~~~~
        359: define void @mutable_unsafe_borrow(ptr noalias noundef align 2 dereferenceable(2) %_1) unnamed_addr #1 { 
check:83     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        360: start: 
check:83     ~~~~~~~
        361:  ret void 
check:83     ~~~~~~~~~~
        362: } 
check:83     ~~
        363:  
check:83     ~
        364: ; Function Attrs: uwtable 
check:83     ~~~~~~~~~~~~~~~~~~~~~~~~~~
        365: define void @mutable_borrow(ptr noalias noundef align 4 dereferenceable(4) %_1) unnamed_addr #1 { 
check:83     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        366: start: 
check:83     ~~~~~~~
        367:  ret void 
check:83     ~~~~~~~~~~
        368: } 
check:83     ~~
        369:  
check:83     ~
        370: ; Function Attrs: uwtable 
check:83     ~~~~~~~~~~~~~~~~~~~~~~~~~~
        371: define noundef align 4 dereferenceable(4) ptr @mutable_borrow_ret() unnamed_addr #1 { 
check:83     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        372: start: 
check:83     ~~~~~~~
        373:  br label %bb1 
check:83     ~~~~~~~~~~~~~~~
        374:  
check:83     ~
        375: bb1: ; preds = %bb1, %start 
check:83     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        376:  br label %bb1 
check:83     ~~~~~~~~~~~~~~~
        377: } 
check:83     ~~
        378:  
check:83     ~
        379: ; Function Attrs: uwtable 
check:83     ~~~~~~~~~~~~~~~~~~~~~~~~~~
        380: define void @mutable_notunpin_borrow(ptr noundef nonnull align 4 %_1) unnamed_addr #1 { 
check:83     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        381: start: 
check:83     ~~~~~~~
        382:  ret void 
check:83     ~~~~~~~~~~
        383: } 
check:83     ~~
        384:  
check:83     ~
        385: ; Function Attrs: uwtable 
check:83     ~~~~~~~~~~~~~~~~~~~~~~~~~~
        386: define void @notunpin_borrow(ptr noalias noundef readonly align 4 dereferenceable(4) %_1) unnamed_addr #1 { 
check:83     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        387: start: 
check:83     ~~~~~~~
        388:  ret void 
check:83     ~~~~~~~~~~
        389: } 
check:83     ~~
        390:  
check:83     ~
        391: ; Function Attrs: uwtable 
check:83     ~~~~~~~~~~~~~~~~~~~~~~~~~~
        392: define void @indirect_struct(ptr noalias nocapture noundef readonly align 4 dereferenceable(32) %_1) unnamed_addr #1 { 
check:83     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        393: start: 
check:83     ~~~~~~~
        394:  ret void 
check:83     ~~~~~~~~~~
        395: } 
check:83     ~~
        396:  
check:83     ~
        397: ; Function Attrs: uwtable 
check:83     ~~~~~~~~~~~~~~~~~~~~~~~~~~
        398: define void @borrowed_struct(ptr noalias noundef readonly align 4 dereferenceable(32) %_1) unnamed_addr #1 { 
check:83     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        399: start: 
check:83     ~~~~~~~
        400:  ret void 
check:83     ~~~~~~~~~~
        401: } 
check:83     ~~
        402:  
check:83     ~
        403: ; Function Attrs: uwtable 
check:83     ~~~~~~~~~~~~~~~~~~~~~~~~~~
        404: define void @option_borrow(ptr noalias noundef readonly align 4 dereferenceable_or_null(4) %_x) unnamed_addr #1 { 
check:83     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        405: start: 
check:83     ~~~~~~~
        406:  ret void 
check:83     ~~~~~~~~~~
        407: } 
check:83     ~~
        408:  
check:83     ~
        409: ; Function Attrs: uwtable 
check:83     ~~~~~~~~~~~~~~~~~~~~~~~~~~
        410: define void @option_borrow_mut(ptr noalias noundef align 4 dereferenceable_or_null(4) %_x) unnamed_addr #1 { 
check:83     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        411: start: 
check:83     ~~~~~~~
        412:  ret void 
check:83     ~~~~~~~~~~
        413: } 
check:83     ~~
        414:  
check:83     ~
        415: ; Function Attrs: uwtable 
check:83     ~~~~~~~~~~~~~~~~~~~~~~~~~~
        416: define void @nonnull_and_nondereferenceable(ptr noundef nonnull %_x) unnamed_addr #1 { 
check:83     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        417: start: 
check:83     ~~~~~~~
        418:  ret void 
check:83     ~~~~~~~~~~
```

</details>

--- a/compiler/rustc_codegen_llvm/src/abi.rs
+++ b/compiler/rustc_codegen_llvm/src/abi.rs
@@ -42,12 +42,13 @@ trait ArgAttributesExt {
 const ABI_AFFECTING_ATTRIBUTES: [(ArgAttribute, llvm::AttributeKind); 1] =
     [(ArgAttribute::InReg, llvm::AttributeKind::InReg)];
 
-const OPTIMIZATION_ATTRIBUTES: [(ArgAttribute, llvm::AttributeKind); 5] = [
+const OPTIMIZATION_ATTRIBUTES: [(ArgAttribute, llvm::AttributeKind); 6] = [
     (ArgAttribute::NoAlias, llvm::AttributeKind::NoAlias),
     (ArgAttribute::NoCapture, llvm::AttributeKind::NoCapture),
     (ArgAttribute::NonNull, llvm::AttributeKind::NonNull),
     (ArgAttribute::ReadOnly, llvm::AttributeKind::ReadOnly),
     (ArgAttribute::NoUndef, llvm::AttributeKind::NoUndef),
+    (ArgAttribute::CapturesReadOnly, llvm::AttributeKind::CapturesReadOnly),
 ];
 
 fn get_attrs<'ll>(this: &ArgAttributes, cx: &CodegenCx<'ll, '_>) -> SmallVec<[&'ll Attribute; 8]> {
@@ -83,6 +84,10 @@ fn get_attrs<'ll>(this: &ArgAttributes, cx: &CodegenCx<'ll, '_>) -> SmallVec<[&'
         }
         for (attr, llattr) in OPTIMIZATION_ATTRIBUTES {
             if regular.contains(attr) {
+                // captures(address, read_provenance) is only available since LLVM 21.
+                if attr == ArgAttribute::CapturesReadOnly && llvm_util::get_version() < (21, 0, 0) {
+                    continue;
+                }
                 attrs.push(llattr.create_attr(cx.llcx));
             }
         }

--- a/compiler/rustc_codegen_llvm/src/llvm/ffi.rs
+++ b/compiler/rustc_codegen_llvm/src/llvm/ffi.rs
@@ -251,6 +251,7 @@ pub(crate) enum AttributeKind {
     Writable = 42,
     DeadOnUnwind = 43,
     DeadOnReturn = 44,
+    CapturesReadOnly = 45,
 }
 
 /// LLVMIntPredicate

--- a/compiler/rustc_llvm/llvm-wrapper/RustWrapper.cpp
+++ b/compiler/rustc_llvm/llvm-wrapper/RustWrapper.cpp
@@ -278,6 +278,7 @@ enum class LLVMRustAttributeKind {
   Writable = 42,
   DeadOnUnwind = 43,
   DeadOnReturn = 44,
+  CapturesReadOnly = 45,
 };
 
 static Attribute::AttrKind fromRust(LLVMRustAttributeKind Kind) {
@@ -376,6 +377,8 @@ static Attribute::AttrKind fromRust(LLVMRustAttributeKind Kind) {
 #else
     report_fatal_error("DeadOnReturn attribute requires LLVM 21 or later");
 #endif
+  case LLVMRustAttributeKind::CapturesReadOnly:
+    report_fatal_error("Should be handled separately");
   }
   report_fatal_error("bad LLVMRustAttributeKind");
 }
@@ -429,6 +432,11 @@ LLVMRustCreateAttrNoValue(LLVMContextRef C, LLVMRustAttributeKind RustAttr) {
   // LLVM 21 replaced the NoCapture attribute with Captures(none).
   if (RustAttr == LLVMRustAttributeKind::NoCapture) {
     return wrap(Attribute::getWithCaptureInfo(*unwrap(C), CaptureInfo::none()));
+  }
+  if (RustAttr == LLVMRustAttributeKind::CapturesReadOnly) {
+    return wrap(Attribute::getWithCaptureInfo(
+        *unwrap(C), CaptureInfo(CaptureComponents::Address |
+                                CaptureComponents::ReadProvenance)));
   }
 #endif
   return wrap(Attribute::get(*unwrap(C), fromRust(RustAttr)));

--- a/compiler/rustc_target/src/callconv/mod.rs
+++ b/compiler/rustc_target/src/callconv/mod.rs
@@ -119,6 +119,7 @@ mod attr_impl {
             const ReadOnly  = 1 << 4;
             const InReg     = 1 << 5;
             const NoUndef = 1 << 6;
+            const CapturesReadOnly = 1 << 7;
         }
     }
     rustc_data_structures::external_bitflags_debug! { ArgAttribute }

--- a/compiler/rustc_ty_utils/src/abi.rs
+++ b/compiler/rustc_ty_utils/src/abi.rs
@@ -356,6 +356,7 @@ fn arg_attrs_for_rust_scalar<'tcx>(
 
             if matches!(kind, PointerKind::SharedRef { frozen: true }) && !is_return {
                 attrs.set(ArgAttribute::ReadOnly);
+                attrs.set(ArgAttribute::CapturesReadOnly);
             }
         }
     }

--- a/tests/codegen-llvm/function-arguments.rs
+++ b/tests/codegen-llvm/function-arguments.rs
@@ -80,7 +80,7 @@ pub fn option_nonzero_int(x: Option<NonZero<u64>>) -> Option<NonZero<u64>> {
     x
 }
 
-// CHECK: @readonly_borrow(ptr noalias noundef readonly align 4 dereferenceable(4) %_1)
+// CHECK: @readonly_borrow(ptr noalias noundef readonly align 4{{( captures\(address, read_provenance\))?}} dereferenceable(4) %_1)
 // FIXME #25759 This should also have `nocapture`
 #[no_mangle]
 pub fn readonly_borrow(_: &i32) {}
@@ -91,12 +91,12 @@ pub fn readonly_borrow_ret() -> &'static i32 {
     loop {}
 }
 
-// CHECK: @static_borrow(ptr noalias noundef readonly align 4 dereferenceable(4) %_1)
+// CHECK: @static_borrow(ptr noalias noundef readonly align 4{{( captures\(address, read_provenance\))?}} dereferenceable(4) %_1)
 // static borrow may be captured
 #[no_mangle]
 pub fn static_borrow(_: &'static i32) {}
 
-// CHECK: @named_borrow(ptr noalias noundef readonly align 4 dereferenceable(4) %_1)
+// CHECK: @named_borrow(ptr noalias noundef readonly align 4{{( captures\(address, read_provenance\))?}} dereferenceable(4) %_1)
 // borrow with named lifetime may be captured
 #[no_mangle]
 pub fn named_borrow<'r>(_: &'r i32) {}
@@ -129,7 +129,7 @@ pub fn mutable_borrow_ret() -> &'static mut i32 {
 // <https://github.com/rust-lang/unsafe-code-guidelines/issues/381>.
 pub fn mutable_notunpin_borrow(_: &mut NotUnpin) {}
 
-// CHECK: @notunpin_borrow(ptr noalias noundef readonly align 4 dereferenceable(4) %_1)
+// CHECK: @notunpin_borrow(ptr noalias noundef readonly align 4{{( captures\(address, read_provenance\))?}} dereferenceable(4) %_1)
 // But `&NotUnpin` behaves perfectly normal.
 #[no_mangle]
 pub fn notunpin_borrow(_: &NotUnpin) {}
@@ -138,12 +138,12 @@ pub fn notunpin_borrow(_: &NotUnpin) {}
 #[no_mangle]
 pub fn indirect_struct(_: S) {}
 
-// CHECK: @borrowed_struct(ptr noalias noundef readonly align 4 dereferenceable(32) %_1)
+// CHECK: @borrowed_struct(ptr noalias noundef readonly align 4{{( captures\(address, read_provenance\))?}} dereferenceable(32) %_1)
 // FIXME #25759 This should also have `nocapture`
 #[no_mangle]
 pub fn borrowed_struct(_: &S) {}
 
-// CHECK: @option_borrow(ptr noalias noundef readonly align 4 dereferenceable_or_null(4) %_x)
+// CHECK: @option_borrow(ptr noalias noundef readonly align 4{{( captures\(address, read_provenance\))?}} dereferenceable_or_null(4) %_x)
 #[no_mangle]
 pub fn option_borrow(_x: Option<&i32>) {}
 
@@ -185,7 +185,7 @@ pub fn _box(x: Box<i32>) -> Box<i32> {
 // With a custom allocator, it should *not* have `noalias`. (See
 // <https://github.com/rust-lang/miri/issues/3341> for why.) The second argument is the allocator,
 // which is a reference here that still carries `noalias` as usual.
-// CHECK: @_box_custom(ptr noundef nonnull align 4 %x.0, ptr noalias noundef nonnull readonly align 1 %x.1)
+// CHECK: @_box_custom(ptr noundef nonnull align 4 %x.0, ptr noalias noundef nonnull readonly align 1{{( captures\(address, read_provenance\))?}} %x.1)
 #[no_mangle]
 pub fn _box_custom(x: Box<i32, &std::alloc::Global>) {
     drop(x)
@@ -208,7 +208,7 @@ pub fn struct_return() -> S {
 #[no_mangle]
 pub fn helper(_: usize) {}
 
-// CHECK: @slice(ptr noalias noundef nonnull readonly align 1 %_1.0, [[USIZE]] noundef %_1.1)
+// CHECK: @slice(ptr noalias noundef nonnull readonly align 1{{( captures\(address, read_provenance\))?}} %_1.0, [[USIZE]] noundef %_1.1)
 // FIXME #25759 This should also have `nocapture`
 #[no_mangle]
 pub fn slice(_: &[u8]) {}
@@ -227,7 +227,7 @@ pub fn unsafe_slice(_: &[UnsafeInner]) {}
 #[no_mangle]
 pub fn raw_slice(_: *const [u8]) {}
 
-// CHECK: @str(ptr noalias noundef nonnull readonly align 1 %_1.0, [[USIZE]] noundef %_1.1)
+// CHECK: @str(ptr noalias noundef nonnull readonly align 1{{( captures\(address, read_provenance\))?}} %_1.0, [[USIZE]] noundef %_1.1)
 // FIXME #25759 This should also have `nocapture`
 #[no_mangle]
 pub fn str(_: &[u8]) {}
@@ -259,7 +259,7 @@ pub fn trait_option(x: Option<Box<dyn Drop + Unpin>>) -> Option<Box<dyn Drop + U
     x
 }
 
-// CHECK: { ptr, [[USIZE]] } @return_slice(ptr noalias noundef nonnull readonly align 2 %x.0, [[USIZE]] noundef %x.1)
+// CHECK: { ptr, [[USIZE]] } @return_slice(ptr noalias noundef nonnull readonly align 2{{( captures\(address, read_provenance\))?}} %x.0, [[USIZE]] noundef %x.1)
 #[no_mangle]
 pub fn return_slice(x: &[u16]) -> &[u16] {
     x

--- a/tests/codegen-llvm/range-attribute.rs
+++ b/tests/codegen-llvm/range-attribute.rs
@@ -67,7 +67,7 @@ pub fn enum2_value(x: Enum2) -> Enum2 {
     x
 }
 
-// CHECK: noundef [[USIZE]] @takes_slice(ptr noalias noundef nonnull readonly align 4 %x.0, [[USIZE]] noundef %x.1)
+// CHECK: noundef [[USIZE]] @takes_slice(ptr {{.*}} %x.0, [[USIZE]] noundef %x.1)
 #[no_mangle]
 pub fn takes_slice(x: &[i32]) -> usize {
     x.len()

--- a/tests/codegen-llvm/read-only-capture-opt.rs
+++ b/tests/codegen-llvm/read-only-capture-opt.rs
@@ -1,0 +1,18 @@
+//@ compile-flags: -C opt-level=3 -Z mir-opt-level=0
+//@ min-llvm-version: 21
+
+#![crate_type = "lib"]
+
+unsafe extern "C" {
+    safe fn do_something(p: &i32);
+}
+
+#[unsafe(no_mangle)]
+pub fn test() -> i32 {
+    // CHECK-LABEL: @test(
+    // CHECK: ret i32 0
+    let i = 0;
+    do_something(&i);
+    do_something(&i);
+    i
+}

--- a/tests/ui/abi/debug.generic.stderr
+++ b/tests/ui/abi/debug.generic.stderr
@@ -939,7 +939,7 @@ error: fn_abi_of(assoc_test) = FnAbi {
                    },
                    mode: Direct(
                        ArgAttributes {
-                           regular: NoAlias | NonNull | ReadOnly | NoUndef,
+                           regular: NoAlias | NonNull | ReadOnly | NoUndef | CapturesReadOnly,
                            arg_ext: None,
                            pointee_size: Size(2 bytes),
                            pointee_align: Some(

--- a/tests/ui/abi/debug.riscv64.stderr
+++ b/tests/ui/abi/debug.riscv64.stderr
@@ -939,7 +939,7 @@ error: fn_abi_of(assoc_test) = FnAbi {
                    },
                    mode: Direct(
                        ArgAttributes {
-                           regular: NoAlias | NonNull | ReadOnly | NoUndef,
+                           regular: NoAlias | NonNull | ReadOnly | NoUndef | CapturesReadOnly,
                            arg_ext: None,
                            pointee_size: Size(2 bytes),
                            pointee_align: Some(

--- a/tests/ui/codegen/equal-pointers-unequal/as-cast/segfault.rs
+++ b/tests/ui/codegen/equal-pointers-unequal/as-cast/segfault.rs
@@ -55,6 +55,7 @@ fn main() {
     // The `Box` has been deallocated by now, so this is a dangling reference!
     let r: &u8 = &*r;
     println!("{:p}", r);
+    println!("{}", i);
 
     // The following might segfault. Or it might not.
     // Depends on the platform semantics

--- a/tests/ui/codegen/equal-pointers-unequal/exposed-provenance/segfault.rs
+++ b/tests/ui/codegen/equal-pointers-unequal/exposed-provenance/segfault.rs
@@ -58,6 +58,7 @@ fn main() {
     // The `Box` has been deallocated by now, so this is a dangling reference!
     let r: &u8 = &*r;
     println!("{:p}", r);
+    println!("{}", i);
 
     // The following might segfault. Or it might not.
     // Depends on the platform semantics

--- a/tests/ui/codegen/equal-pointers-unequal/strict-provenance/segfault.rs
+++ b/tests/ui/codegen/equal-pointers-unequal/strict-provenance/segfault.rs
@@ -58,6 +58,7 @@ fn main() {
     // The `Box` has been deallocated by now, so this is a dangling reference!
     let r: &u8 = &*r;
     println!("{:p}", r);
+    println!("{}", i);
 
     // The following might segfault. Or it might not.
     // Depends on the platform semantics

